Fix build for PPC.

shenhanc78 opened the pull request:
This CL includes 2 parts -
1) - provided a speciliazation for vector types on non-sse
platforms,
the default implementation does not compile for Packet4f type.
2) - fixed the default implementation (which is unreachable) for 2
functions, the origin version does not compile for Packet4f type.

Tested:
Building on ppc machines and x86 machines.

shenhanc78 committed:
Fix build for PPC.

This CL includes 2 part -
1) - provided a speciliazation for vector types on non-sse
platforms,
the default implementation does not compile for Packet4f type.
2) - fixed the default implementation (which is unreachable) for 2
functions, the origin version does not compile for Packet4f type.

Tested:
Building on ppc machines and x86 machines.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

benoitsteiner commented on tensorflow/core/kernels/sparse_matmul_op.h:
```diff
@@ -40,6 +40,36 @@ EIGEN_DEVICE_FUNC inline Packet pexpand_bf16_u(const Packet& from) {
+#ifndef EIGEN_VECTORIZE_SSE2
+template <typename Packet>
+EIGEN_DEVICE_FUNC inline Packet4f pexpand_bf16_l(const Packet4f& from) {
+  Packet4f res;
```
The packet4f struct should remain opaque. A more portable version of the code would be:
float r[4];
pstoreu(r, from);
float p[4];
 p[0] = static_cast<float>((r[2] << 16) & 0xffff0000);
 p[1] = static_cast<float>(r[2] & 0xffff0000);
 p[2] = static_cast<float>((r[3] << 16) & 0xffff0000);
 p[3] = static_cast<float>(r[3] & 0xffff0000);
return ploadu(p);

shenhanc78 replied:
Hi Benoit, thanks for the review. Fixed. Uploaded another CL. 

Han

shenhanc78 committed:
Keep type "packet4f struct" opaque.

Tested: build on ppc and x86 machines.

benoitsteiner commented:
Jenkins, test this please.

shenhanc78 committed:
Fixed runtime errors (exposed by tests) and provide more specialization.

In CL2 - "static_cast<int>(a_float)" is not the proper way to view the
underlying memory as integer, it should be re-written as
"reinterpret_cast<int *>(&a_float)".

This CL (CL3) also add speicialization for pload2bf16 and pload4bf16 on non_sse2
platform.

Tested:
Build test - cross build for ppc and native build on x86_64. (via blaze build)

Runtime test - tested by replacing original sse2 version with my
non-sse2 version, passed x86_64 test "sparse_matmul_test" (via blaze
test). Though I didn't run on ppc machines, because we still have other setup
problems.

shenhanc78 commented:
Hi Benoit & Jenkins, thanks. Please test with CL3 (e1bdac5), I ran a local test (via blaze) for CL2, it exposed a few runtime errors, which I fixed in CL3. More detail in CL3's comment.

Thanks,

benoitsteiner commented:
Jenkins, test this please.
